### ISSUE TITLE
Flutter should warn you on Catalina if you're about to use an unsigned binary

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#40905]
UPDATE: It appears there are situations (https://github.com/flutter/flutter/issues/42302 & https://github.com/flutter/flutter/issues/42714) where Dart is allowed to run but libimobiledevice binaries fail. We should probably add a doctor validator to verify our binaries will actually run.

We have plans to sign binaries for Catalina, but there will be binaries from some commits that are not signed (e.g. commits from before we started signing at all, or currently commits that fall between dev tagged branches and aren't getting signed). We should try to make sure that experience is accounted for.

One option is for the Flutter tool (perhaps in the `flutter` bash script) to do something like the following:

If we're on Catalina:
- If `spctl` is on the path, use it to check if `dart` and `flutter_tester` are signed. If not, print a warning/link on how to override the setting on Catalina (or how to get on a channel that is signed).
- If `spctl` is not available, print a warning that we couldn't find the Xcode commandline tools and a link about how to install them, and a warning that the user may see other warnings about `dart` or `flutter_tester`, and a link about how to resolve those.

/cc @christopherfujino @zanderso 
/cc @devoncarew @jacob314 @DanTup  for input on how this might affect IDEs on Catalina.

**Comment by christopherfujino** [comment]
We should also think about do we care if the binaries are signed if they've already disabled Gatekeeper?

**Comment by dnfield** [comment]
Is there a way to check that (without sudo)?  If so we could just do that first and short-circuit everything else.

**Comment by christopherfujino** [comment]
I'll investigate that

**Comment by christopherfujino** [comment]
`spctl --status` will report if Gatekeeper is enabled or not. We definitely should short circuit if they've disabled it.

**Comment by DanTup** [comment]
> /cc @DanTup for input on how this might affect IDEs on Catalina.

I'm not very familiar with this issue - is the problem that running the Dart VM (to run the flutter_tools snapshot) will fail?

For `flutter run`, as long as the shell script terminates nicely with an error, I think that would show up in the debug console to front of the user (so if it has a good message, it should be easy for them to resolve).

For other invocations of Flutter (like `flutter daemon`) we might need to make sure these are handled (I suspect VS code would trigger the "flutter daemon has terminated" message, but I'm not certain).

Is there a way to mimic the same behaviour on High Sierra to test? If not, looks like Catalina isn't far off, so once it lands I can upgrade my personal MacBook and test some there.

**Comment by eseidelGoogle** [comment]
I believe Gatekeeper is always on these days.  Our setup instructions make no mention of if as far as I can tell:
https://flutter.dev/docs/get-started/install/macos
But users get confused by it:
https://twitter.com/koehntopp/status/1387071996001984515

The english version is: “Dart cannot be opened because the developer cannot be verified”  “Move to trash?”

**Comment by eseidelGoogle** [comment]
https://github.com/flutter/flutter/issues/36714 says this was fixed in Dart, but I definitely hit this when attempting to get my wife set up with Flutter.  I'm pretty sure we were installing "stable" as of March 29, 2021 (I can check on her computer if needed).

**Comment by christopherfujino** [comment]
@eseidelGoogle do you know which binary Gatekeeper is flagging? I checked stable 2.0.5 and it looks like all the macos cache binaries are codesigned and notarized.

**Comment by eseidelGoogle** [comment]
In both the above tweet and my wife's experience the binary mentioned in the dialog was titled `dart`.  I don't know if that's the script called Dart or an executable though.

![image](https://user-images.githubusercontent.com/11857803/116301830-5381b400-a755-11eb-9e4b-7c8360a967bc.png)

**Comment by eseidelGoogle** [comment]
I just ran `flutter doctor` on my wife's machine and got:

```
% flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[!] Flutter (Channel stable, 2.0.3, on macOS 11.2.1 20D74 darwin-x64, locale en-US)
    ✗ Downloaded executables cannot execute on host.
      See https://github.com/flutter/flutter/issues/6207 for more information
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.
[✗] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ✗ CocoaPods not installed.
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on
        the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
[✓] Chrome - develop for the web
[!] Android Studio (not installed)
[✓] VS Code (version 1.55.0)
[✓] Connected device (1 available)
! Doctor found issues in 4 categories.
```

We never actually got through setup all the way (as is obvious above).  When I ran `flutter doctor` just now I immediately got a gatekeeper popup:
<img width="372" alt="Screen Shot 2021-04-27 at 12 41 38 PM" src="https://user-images.githubusercontent.com/11857803/116302847-88423b00-a756-11eb-950b-f910ca4fb169.png">

I'm happy to file a separate issue. It's also possible this was only an issue in 2.0.3 and has since been fixed?

**Comment by christopherfujino** [comment]
> In both the above tweet and my wife's experience the binary mentioned in the dialog was titled `dart`. I don't know if that's the script called Dart or an executable though.
> 
> ![image](https://user-images.githubusercontent.com/11857803/116301830-5381b400-a755-11eb-9e4b-7c8360a967bc.png)

I don't know German, but it looks like it's saying Dart was downloaded via Homebrew Cask? Not sure how that happened (unless you have homebrew dart configured via the IDE plugin...)

**Comment by christopherfujino** [comment]
> I just ran `flutter doctor` on my wife's machine and got:
> 
> ```
> % flutter doctor
> Doctor summary (to see all details, run flutter doctor -v):
> [!] Flutter (Channel stable, 2.0.3, on macOS 11.2.1 20D74 darwin-x64, locale en-US)
>     ✗ Downloaded executables cannot execute on host.
>       See https://github.com/flutter/flutter/issues/6207 for more information
> [✗] Android toolchain - develop for Android devices
>     ✗ Unable to locate Android SDK.
>       Install Android Studio from: https://developer.android.com/studio/index.html
>       On first launch it will assist you in installing the Android SDK components.
>       (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
>       If the Android SDK has been installed to a custom location, please use
>       `flutter config --android-sdk` to update to that location.
> [✗] Xcode - develop for iOS and macOS
>     ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
>       Download at: https://developer.apple.com/xcode/download/
>       Or install Xcode via the App Store.
>       Once installed, run:
>         sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
>         sudo xcodebuild -runFirstLaunch
>     ✗ CocoaPods not installed.
>         CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on
>         the Dart side.
>         Without CocoaPods, plugins will not work on iOS or macOS.
>         For more info, see https://flutter.dev/platform-plugins
>       To install see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.
> [✓] Chrome - develop for the web
> [!] Android Studio (not installed)
> [✓] VS Code (version 1.55.0)
> [✓] Connected device (1 available)
> ! Doctor found issues in 4 categories.
> ```
> 
> We never actually got through setup all the way (as is obvious above). When I ran `flutter doctor` just now I immediately got a gatekeeper popup:
> <img alt="Screen Shot 2021-04-27 at 12 41 38 PM" width="372" src="https://user-images.githubusercontent.com/11857803/116302847-88423b00-a756-11eb-950b-f910ca4fb169.png">
> 
> I'm happy to file a separate issue. It's also possible this was only an issue in 2.0.3 and has since been fixed?

Yeah, this should be a different issue. Feel free to ping me on it, `gen_snapshot` from the website is our fault.

**Comment by eseidelGoogle** [comment]
Filed https://github.com/flutter/flutter/issues/81326

**Comment by dnfield** [comment]
I think we can close this - Catalina is no longer the most recent version, and AFAIK running binaries on recent versions of macOS that come from GitHub non-stable aren't causing problems.
